src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -256,3 +266,227 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+                                                      uint8_t* record_header,
+                                                      char** error_details) {
+  GPR_ASSERT(record_header != nullptr);
+  if (header_size != SSL3_RT_HEADER_LENGTH) {
+    *error_details = gpr_strdup(
+        "The header size does not match the size of a TLS 1.3 record header.");
+    return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+  record_header[0] = record_type;
+  const uint16_t wire_version = static_cast<uint16_t>(TLS1_2_VERSION);
+  record_header[1] = wire_version >> 8;
+  record_header[2] = wire_version & 0xff;
+  record_header[3] = payload_size >> 8;
+  record_header[4] = payload_size & 0xff;
+  return GRPC_STATUS_OK;
+}
+
+/** This method returns a buffer of additional data bytes used for
+ *  authenticating a TLS 1.3 record; the caller must free the buffer using
+ *  gpr_free.
+ *  - sequence: a buffer containing the sequence number of a half connection.
+ *  - sequence_size: the size (in bytes) of the |sequence| buffer.
+ *  - record_header: a buffer containing the header of a TLS 1.3 record.
+ *  - header_size: the size (in bytes) of the |record_header| buffer.
+ *  - payload_size: the size (in bytes) of the payload of a TLS 1.3 record. **/
+static uint8_t* s2a_additional_data(uint8_t* sequence, size_t sequence_size,
+                                    uint8_t* record_header, size_t header_size,
+                                    size_t payload_size) {
+  size_t additional_data_size = sequence_size + header_size;
+  GPR_ASSERT(additional_data_size == kTlsAdditionalDataBytesSize);
+  uint8_t* additional_data =
+      static_cast<uint8_t*>(gpr_malloc(additional_data_size * sizeof(uint8_t)));
+  memcpy(additional_data, sequence, sequence_size);
+  memcpy(additional_data + sequence_size, record_header, header_size);
+  additional_data[11] = payload_size >> 8;
+  additional_data[12] = payload_size & 0xff;
+  return additional_data;
+}
+
+/** This method returns a buffer populated with the nonce used for
+ *  encrypting and decrypting a TLS 1.3 record; the caller must free
+ *  the buffer using gpr_free.
+ *  - crypter: an instance of s2a_crypter.
+ *  - sequence: a buffer containing the sequence number of a half connection.
+ *  - sequence_size: the size (in bytes) of the |sequence| buffer.
+ *  - nonce_size: the number of bytes written to the nonce buffer that is
+ *    returned by the method. **/
+static uint8_t* s2a_nonce(s2a_crypter* crypter, uint8_t* sequence,
+                          size_t sequence_size, size_t* nonce_size) {
+  GPR_ASSERT(crypter != nullptr);
+  size_t max_nonce_size = s2a_max_aead_nonce_size(crypter);
+  GPR_ASSERT(max_nonce_size > sequence_size);
+  GPR_ASSERT(sequence != nullptr);
+  GPR_ASSERT(nonce_size != nullptr);
+  uint8_t* nonce =
+      static_cast<uint8_t*>(gpr_malloc(max_nonce_size * sizeof(uint8_t)));
+  *nonce_size = crypter->out_connection->fixed_nonce_size;
```

Comment:
How about adding GPR_ASSERT for `crypter->out_connection != nullptr`?

Reply:
Good call. Done.